### ISSUE TITLE
AArch64 macOS: Failure in JDK 11 java/lang/invoke/VarHandles/VarHandleTestByteArrayAsInt

### ISSUE THREAD

**Issue by knn-k** [eclipse-openj9/openj9#14636]
Failure link
------------
https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.openjdk_aarch64_mac_Nightly/2/

Optional info
-------------

- intermittent failure: yes
- regression or new test: regression

Failure output (captured from console output)
---------------------------------------------
No output available: Only `Unexpected exit from test [exit code: 255]`
https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.openjdk_aarch64_mac_Nightly/2/testReport/junit/java_lang_invoke_VarHandles_VarHandleTestByteArrayAsInt/java/VarHandleTestByteArrayAsInt/

**Comment by pshipton** [comment]
May be related to https://github.com/eclipse-openj9/openj9/issues/13368

**Comment by pshipton** [comment]
Ignore the previous comment, there is additional output

```
23:22:37  04:22:16.257 0x12606f100    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK11_aarch64_mac_Nightly/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
```

**Comment by pshipton** [comment]
Also artifacts https://openj9-artifactory.osuosl.org/artifactory/ci-openj9/Test/Test_openjdk11_j9_sanity.openjdk_aarch64_mac_Nightly/2/openjdk_test_output.tar.gz

**Comment by knn-k** [comment]
I ran Grinder jobs below, and they were successful.
- https://openj9-jenkins.osuosl.org/job/Grinder/689/
- https://openj9-jenkins.osuosl.org/job/Grinder/690/

**Comment by pshipton** [comment]
https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.system_aarch64_mac_Nightly_testList_0/3 - [mac11-aarch64-8](https://openj9-jenkins.osuosl.org/computer/mac11-aarch64-8)
LambdaLoadTest_J9_5m_0 `-Xjit -Xgcpolicy:gencon -Xnocompressedrefs`

https://openj9-artifactory.osuosl.org/artifactory/ci-openj9/Test/Test_openjdk11_j9_sanity.system_aarch64_mac_Nightly_testList_0/3/system_test_output.tar.gz
```
LT  stderr 05:52:26.178 0x142820700    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK11_aarch64_mac_Nightly/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
LT  stderr 05:52:26.178 0x142837700    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK11_aarch64_mac_Nightly/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
LT  stderr 05:52:26.178 0x14280af00    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK11_aarch64_mac_Nightly/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
LT  stderr 05:52:26.178 0x153193500    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK11_aarch64_mac_Nightly/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
LT  stderr 05:52:26.178 0x142814f00    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK11_aarch64_mac_Nightly/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
LT  stderr 05:52:26.178 0x1530c0900    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK11_aarch64_mac_Nightly/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
LT  stderr 05:52:26.178 0x1530c9900    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK11_aarch64_mac_Nightly/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
```

**Comment by knn-k** [comment]
20x Grinder job for LambdaLoadTest_J9_5m_0 did not fail: internal job/Grinder/21639/

**Comment by knn-k** [comment]
Possible causes:
- `MM_ParallelTask::synchronizeGCThreadsAndReleaseSingleThread()` finishes too early
- Values returned from `omrtime_hires_clock()` wrong or imprecise
- `_syncCriticalSectionDuration` has unexpected value

**Comment by knn-k** [comment]
Grinder jobs for jdk_lang_1 did not fail: job/Grinder/21661/, job/Grinder/21665/

**Comment by knn-k** [comment]
Reproduced in job/Test_openjdk18_j9_special.system_aarch64_mac_Personal_testList_4/1/consoleText in running DaaLoadTest_all_special_5m_9.
```
[2022-04-05T05:07:52.959Z] DLT 01:07:52.688 - Starting thread. Suite=0 thread=5
[2022-04-05T05:08:01.245Z] DLT stderr 05:07:59.887 0x15996ab00    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK18_aarch64_mac_Personal/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
[2022-04-05T05:08:01.245Z] DLT stderr 05:07:59.887 0x159999500    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK18_aarch64_mac_Personal/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
```

**Comment by knn-k** [comment]
I ran a 20x Grinder job for DaaLoadTest_all_special_5m_9 in job/Grinder/22364/, and got no failures.

**Comment by pshipton** [comment]
https://openj9-jenkins.osuosl.org/job/Test_openjdk18_j9_special.system_aarch64_mac_Personal_testList_0/2
DaaLoadTest_daa2_special_5m_29 `-Xgcpolicy:gencon -Xgc:dynamicBreadthFirstScanOrdering`

```
DLT 06:03:08.407 - Completed 26.7%. Number of tests started=35820 (+11485)
DLT stderr 20:03:13.757 0x14e106900    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK18_aarch64_mac_Personal/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
```

**Comment by dmitripivkine** [comment]
Sorry, I missed there is assertion in GC code

This is snap traces:
```
20:03:13.757515000        *0x14e106900       j9mm.542  Entry      >Trc_MM_SynchronizeGCThreadsAndReleaseSingleThread Entry from: /Users/jenkins/workspace/Build_JDK18_aarch64_mac_Personal/openj9/runtime/gc_glue_java/ScavengerRootScanner.hpp:168
20:03:13.757528000        *0x14e91e900       j9mm.542  Entry      >Trc_MM_SynchronizeGCThreadsAndReleaseSingleThread Entry from: /Users/jenkins/workspace/Build_JDK18_aarch64_mac_Personal/openj9/runtime/gc_glue_java/ScavengerRootScanner.hpp:168
20:03:13.757529000         0x14e91e900       j9mm.543  Exit       <Trc_MM_SynchronizeGCThreadsAndReleaseSingleThread Exit
20:03:13.757555000         0x14e91e900       j9mm.122  Entry      >Trc_MM_SynchronizeGCThreads Entry from: /Users/jenkins/workspace/Build_JDK18_aarch64_mac_Personal/openj9/runtime/gc_glue_java/ScavengerRootClearer.hpp:118
20:03:13.757561000        *0x14e106900       j9mm.107  Assert    * ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK18_aarch64_mac_Personal/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
20:03:13.757561000         0x14e106900       j9mm.543  Exit       <Trc_MM_SynchronizeGCThreadsAndReleaseSingleThread Exit
```
I believe last two lines are in incorrect order due both have the same time stamp

There are two threads perform GC operation:
- Thread `0x14e106900` reached `ScavengerRootScanner.hpp:168` and waited for `0x14e91e900` in `MM_ParallelTask::synchronizeGCThreadsAndReleaseSingleThread()`. 
- Thread `0x14e91e900` arrived, saved `_syncCriticalSectionStartTime` and execute settings of flags in `MM_ScavengerRootScanner::scanClearable()` single threaded. After this it called `releaseSynchronizedGCThreads()` where `_syncCriticalSectionDuration` is calculated. 
- Thread `0x14e106900` was waiting on monitor all this time and was released at the end of `releaseSynchronizedGCThreads()` by thread `0x14e91e900`. So, time between start and end in thread `0x14e106900` must be larger or equal that `_syncCriticalSectionDuration`. 
- Thread `0x14e91e900` meanwhile reached next sync point at `ScavengerRootClearer.hpp:118` and was waiting for thread `0x14e106900` there.

As was stated correctly in the comment https://github.com/eclipse-openj9/openj9/issues/14636#issuecomment-1062567347
the reason for assertion might be different flow of operations (monitor does not work properly?) or inconsistent readings of wall clock (?). When we will have DDR support available we can check `_syncCriticalSectionStartTime` and `_syncCriticalSectionDuration`. Also in theory it is possible there is might be some compiler side effect (`_syncCriticalSectionDuration` is not declared as `volatile` - I think it might need be). We can check generated sequence of instructions (but if it is a case I would expect larger number of failures)

**Comment by knn-k** [comment]
DDR implementation for AArch64 macOS is in progress in https://github.com/eclipse/omr/pull/6471 and #14921.

**Comment by pshipton** [comment]
amac DDR is now enabled in future builds.

**Comment by pshipton** [comment]
https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_special.system_aarch64_mac_Release_testList_4/3 - [mac11-aarch64-6](https://openj9-jenkins.osuosl.org/computer/mac11-aarch64-6)
DaaLoadTest_daa3_special_5m_5 `-Xgcpolicy:gencon -Xjit:count=0,optlevel=warm,gcOnResolve,rtResolve -Xmn512k -Xnocompressedrefs`

https://openj9-artifactory.osuosl.org/artifactory/ci-openj9/Test/Test_openjdk17_j9_special.system_aarch64_mac_Release_testList_4/3/system_test_output.tar.gz

```
DLT 23:13:00.903 - Completed 80.1%. Number of tests started=12005 (+1002)
DLT stderr 13:13:03.186 0x13901b300    j9mm.107    *   ** ASSERTION FAILED ** at /Users/jenkins/workspace/Build_JDK17_aarch64_mac_Release/omr/gc/base/standard/ParallelScavengeTask.cpp:120: ((false && ((endTime - startTime) >= _syncCriticalSectionDuration)))
```
@dmitripivkine fyi - DDR is enabled now.

**Comment by dmitripivkine** [comment]
This assertion does check that waiting time in the `synchronizeGCThreadsAndReleaseSingleThread()` for blocked thread should be not smaller than total execution time reordered by thread performed single threaded operation. However total execution time reordered by thread performed single threaded operation is negative (-11 ticks). And treated as unsigned `0xFFFFFFFFFFFFFFE7` it happen to be larger than waiting time of blocked thread

**Comment by dmitripivkine** [comment]
The execution time for thread performed single threaded operation was very short in this case. I have not looked in c-stack/registers for values in ticks but Snap traces shows 0us difference, so it was under 1us. And somehow EndTime (taken at the end) is smaller by 25 ticks as StartTime(taken at the beginning). Can actually wall clock adjustment (backwards) be visible by one thread and missed by another?

**Comment by dmitripivkine** [comment]
In order to get time GC code uses `omrtime_hires_clock()`. OSX uses general Unix implementation based on `gettimeofday()`. This is reading of wall clock and it is not monotonic. So, it can jump back and force.
There is old but I believe still valid blog on topic:
https://web.archive.org/web/20100517095152/http://www.wand.net.nz/~smr26/wordpress/2009/01/19/monotonic-time-in-mac-os-x/comment-page-1/

**Comment by knn-k** [comment]
I wonder why the failure happens only on AArch64 macOS if it is caused by lacking the monotonicity of the clock.
Potentially x86 macOS has the same issue, but M1 chip is so fast that it reveals the issue?

**Comment by knn-k** [comment]
I opened https://github.com/eclipse/omr/pull/6568 for monotonic `omrtime_hires_clock()` implementation on macOS.
I ran sanity.functional and sanity.openjdk with it, and the tests pass.

- sanity.functional: job/Test_openjdk11_j9_sanity.functional_aarch64_mac_Personal/21/
- sanity.openjdk: job/Grinder/24972/

**Comment by knn-k** [comment]
https://github.com/eclipse/omr/pull/6568 was superseded by https://github.com/eclipse/omr/pull/6573.
I use nanoseconds as the unit for `omrtime_hires_clock()` on macOS in it.

**Comment by knn-k** [comment]
OMR PR 6573 was merged.  Let's see if the assertion failure is reproduced or not with the change.

**Comment by knn-k** [comment]
It is a week since OMR PR 6573 was merged, and I don't think we have seen a new failure.
Shall I open another PR for v0.33.0?

**Comment by knn-k** [comment]
I opened https://github.com/eclipse-openj9/openj9-omr/pull/150 for v0.33.0.  It is in the draft state.

**Comment by pshipton** [comment]
I've merged it.
